### PR TITLE
Megafauna Boss Music

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request

Ports https://github.com/yogstation13/Yogstation/pull/11024
Changes the original famitracker chiptunes to use instruments instead, hopefully giving each fight an unique feel
Example:
https://streamable.com/7kzlie
https://streamable.com/nhmgi8
https://streamable.com/2cdwbs
Themes:
[Ash Drake/Clockwork Defender](https://vocaroo.com/15vyaixzIsJd)
[Bubblegum](https://vocaroo.com/1mWX2new1Ioe)
[Colossus](https://vocaroo.com/1obiw3BE9c1a)
[Hierophant/Swarmer Beacon](https://voca.ro/15KKd38MuUD9)
[Blood Drunk Miner/Demonic Frost Miner](https://vocaroo.com/1dqE0xhzxk0k)
[Wendigo](https://vocaroo.com/1nK8z3RED4kS)

## Why It's Good For The Game

It gives the fights life, makes them feel more like real bosses.

## Changelog

:cl: monster860, Fikou
add: Megafauna Boss Music
/:cl:

### CODE REVIEW

File: code/datums/music.dm [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+	var/fade_in_time = 30
+	/// Whether music of the same typepath is shared
+	var/shared = TRUE
+
+/datum/component/music_player/Initialize(_music_path)
+	if(_music_path)
+		music_path = _music_path
+	START_PROCESSING(SSprocessing, src)
+
+/datum/component/music_player/proc/do_range_check(var/fade_time = fade_in_time)
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:198]: hey?

File: code/datums/music.dm [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	if(!is_fading || fade_volume_target != target)
+		fade_at_rate(target, range_fade_speed)
+
+	. = ..()
+
+/datum/music/sourced/Destroy()
+	. = ..()
+	for(var/_player in players)
+		var/datum/component/music_player/player = _player
+		player.mob_players -= target_mob
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:180]:
```suggestion
	for(var/datum/component/music_player/player as anything in players)
		player.mob_players -= target_mob
```

File: code/datums/music.dm [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+/datum/component/music_player
+	var/list/mob_players = list()
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:183]: undocumented

File: code/datums/music.dm [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+	var/music_path
+	/// Whether new mobs will be added. To force mobs to stop playing this, use stop_all()
+	var/enabled = TRUE
+	/// Range within which music will start to play
+	var/start_range = 7
+	var/fade_in_time = 30
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:189]: undocumented

File: code/datums/music.dm [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	var/shared = FALSE
+	for(var/mob/living/M in range(start_range, parent))
+		if(!M.client)
+			continue
+		if(mob_players[M])
+			continue
+		var/did_find = FALSE
+		if(shared)
+			for(var/_music in M.client.active_music)
+				var/datum/music/sourced/music = _music
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:210]:
```suggestion
			for(var/datum/music/sourced/music as anything in M.client.active_music)
```

File: code/datums/music.dm [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+
+/datum/component/music_player/proc/stop_all(fade_time = 0)
+	for(var/_M in mob_players)
+		var/datum/music/sourced/music = mob_players[_M]
+		music.fade(0, fade_time)
+
+/datum/component/music_player/proc/remove_all()
+	for(var/_M in mob_players)
+		var/mob/M = _M
```
Comment (JohnFulpWillard) [inline code/datums/music.dm:232]: im gonna stop repeating myself

**Comment by DragonTrance** [comment]
Does this actually work in fading out music when the megafauna doesn't see you? I didn't think byond was capable of fading out music

**Comment by Fikou** [comment]
> Does this actually work in fading out music when the megafauna doesn't see you (or you leave the area)? I didn't think byond was capable of fading out music

yes

**Comment by NamelessFairy** [comment]
Very much not a fan of the instrument choices and I think the original chiptune themes would be far preferred. The blood drunk miner themes stands out as the worst offender with a constant droning noise that plays throughout that overshadows the actual song.

**Comment by Livrah** [comment]
"Why Do I Hear a Boss Music"

**Comment by monster860** [comment]
*disclaimer: I do not play on TG*

As the one who wrote the original music I will offer my thoughts:

There's a few problems I've noticed. The first problem is that there's a lot of clipping. This is clear just from downloading the files and plonking them into Audacity and looking. There's also the fact that a lot of the instruments chosen are quite muddy. For example if you listen to the original for the "blood drunk miner" (king goat originally), [found here](https://github.com/yogstation13/Yogstation/blob/master/sound/lavaland/music/kilogram.ogg), the beginning there is a lot of emphasis on the 8th notes, as well as the syncopated eight notes. However, in this new version, They are barely audible compared to... some kind of low droning. In addition there's things straight up missing. 4 measures in, in the original, there is a percussion bit that comes in, and I can't hear that at all in this new version.

There's more missing stuff I guess. [as you can hear in this comparison](https://cdn.discordapp.com/attachments/926909577105117235/926915852392104046/heirophant_comparison.ogg), in the hierophant music there's a high square-wave bit that crescendos in and decrescendos out, that's completely gone in this new version.

There's also a general lack of consideration into relative loudness. For example, the beginning of the collossus theme seems barely audible in the new one.

There's probably more stuff, but that's my comments for now.

**Comment by Fikou** [comment]
> _disclaimer: I do not play on TG_
> 
> As the one who wrote the original music I will offer my thoughts:
> 
> There's a few problems I've noticed. The first problem is that there's a lot of clipping. This is clear just from downloading the files and plonking them into Audacity and looking. There's also the fact that a lot of the instruments chosen are quite muddy. For example if you listen to the original for the "blood drunk miner" (king goat originally), [found here](https://github.com/yogstation13/Yogstation/blob/master/sound/lavaland/music/kilogram.ogg), the beginning there is a lot of emphasis on the 8th notes, as well as the syncopated eight notes. However, in this new version, They are barely audible compared to... some kind of low droning. In addition there's things straight up missing. 4 measures in, in the original, there is a percussion bit that comes in, and I can't hear that at all in this new version.
> 
> There's more missing stuff I guess. [as you can hear in this comparison](https://cdn.discordapp.com/attachments/926909577105117235/926915852392104046/heirophant_comparison.ogg), in the hierophant music there's a high square-wave bit that crescendos in and decrescendos out, that's completely gone in this new version.
> 
> There's also a general lack of consideration into relative loudness. For example, the beginning of the collossus theme seems barely audible in the new one.
> 
> There's probably more stuff, but that's my comments for now.

fair point, im not a music guy, just thought chiptunes dont really fit the vibes at tg, ingame it seems fine but ill look more

**Comment by MrDoomBringer** [comment]
chiptunes definitely fit the vibe

**Comment by D4C-420** [comment]
There is a way to turn off the music right? Boss music is cool and all but sometimes you want to hear the environment and boss more than the ethereal band of clowns playing from space.

**Comment by Fikou** [comment]
> There is a way to turn off the music right? Boss music is cool and all but sometimes you want to hear the environment and boss more than the ethereal band of clowns playing from space.

yes its handled by instruments prefs

**Comment by Shadowflame909** [comment]
When the orchestral music starts on lavaland 😱

**Comment by Fikou** [comment]
i guess this is wip with the tracks

**Pull request closed without merging**